packit_service/service/models.py

```diff
@@ -0,0 +1,174 @@
+"""
+Data models for jobs, tasks, builds, etc. The models are mapped to redis.
+"""
+import copy
+import datetime
+from typing import List, Dict, Union
+
+from frambo.dict_in_redis import PersistentDict
+
+from packit_service.service.events import InstallationEvent, WhitelistStatus
+
+
+class Model:
+    """ Abstract representation of a single object passing through p-s """
+
+    # a "table" to store collection of objects
+    table_name: str
+    # unique identifier
+    identifier: Union[int, str] = None
+
+    @classmethod
+    def db(cls) -> PersistentDict:
+        if not cls.table_name:
+            raise RuntimeError("table_name is not set")
+        return PersistentDict(hash_name=cls.table_name)
+
+    def save(self):
+        """ store the current state of the object inside redis """
+        db = self.db()
+        if not self.identifier:
+            last_index = db["last_index"]
+            if last_index is None:
+                last_index = 0
+            self.identifier = last_index + 1
+            db["last_index"] = self.identifier
+        self.db()[self.identifier] = self.serialize()
+
+    def serialize(self):
+        """ convert from python data structure into format acceptable by redis """
+        data = self.__dict__
+        cp = copy.deepcopy(data)
+        if "table_name" in cp:
+            del cp["table_name"]
+        return cp
+
+    def deserialize(self, inp: Dict):
+        """ reverse operation as serialize """
+        # this is pretty nasty: we could possibly replace this with a serialization
+        # library or an ORM framework
```

Comment:
+1 for research of alternatives

Reply:
I did not want to use any library here since it would make this PR gigantic

I also think that it should be fairly easy to migrate to a proper ORM solution in future